File: lib/raptor-server/handler.rb
# Handler class.
# Handles a client connected via a websocket connection.

require 'active_support/core_ext/hash'
require 'securerandom'
require 'signature'
require 'fiber'

module RaptorServer
  class Handler

    attr_accessor :connection
    delegate :error, :send_payload, to: :connection

    def initialize(socket, handshake)
      @socket        = socket
      @handshake = handshake
      @connection    = Connection.new(@socket)
      @subscriptions = {}
      authenticate
    end

    # Dispatches message handling to method with same name as
    # the event name
    def onmessage(msg)
      msg   = JSON.parse msg
      event = msg['event'].gsub(/^\/raptor\//, 'raptor_')

      if event =~ /^\/client\//
        msg['socket_id'] = connection.socket_id
        Channel.send_client_message msg
      elsif respond_to? event, true
        send event, msg
      end

    rescue JSON::ParserError
      error({ code: 5001, message: "Invalid JSON" })
    rescue Exception => e
      error({ code: 500, message: "#{e.message}\n #{e.backtrace.join "\n"}" })
    end

    def onclose
      @subscriptions.select { |k,v| k && v }.
        each do |channel_id, subscription_id|
          Channel.unsubscribe channel_id, subscription_id
        end
    end

    def authenticate
      return connection.establish if valid_app_key? app_key

      error({ code: 4001, message: "Could not find app by key #{app_key}" })
      @socket.close_websocket
    end

    def raptor_ping(msg)
      send_payload nil, '/raptor/pong'
    end

    def raptor_pong msg; end

    def raptor_subscribe(msg)
      channel_id = msg['data']['channel']
      klass      = subscription_klass channel_id

      if @subscriptions[channel_id]
        error({ code: nil, message: "Existing subscription to #{channel_id}" })
      else
        @subscriptions[channel_id] = klass.new(connection.socket, connection.socket_id, msg).subscribe
      end
    end

    def raptor_unsubscribe(msg)
      channel_id      = msg['data']['channel']
      subscription_id = @subscriptions.delete(channel_id)

      Channel.unsubscribe channel_id, subscription_id
    end

    private

    def app_key
      @handshake.path.split('/')[1]
    end

    def valid_app_key? app_key
      RaptorServer::Config.app_key == app_key
    end

    def subscription_klass channel_id
      klass = channel_id.match(/^\/(private|presence)\//) do |match|
        RaptorServer.const_get "#{match[1]}_subscription".classify
      end

      klass || RaptorServer::Subscription
    end
  end
end

File: lib/handler.js
var urlParser = require('url');
var Connection = require('./connection');
var _ = require('lodash');

function Handler(socket, handshake) {
  this.socket        = socket;
  this.handshake     = handshake;
  this.connection    = new Connection(this.socket, null);
  this.subscriptions = {};
  this.authenticate();
}

// Delegates
Handler.prototype.error = function error(e) {
  this.connection.error(e);
};

Handler.prototype.send_payload = function send_payload(channel_id, event_name, payload) {
  this.connection.send_payload(channel_id, event_name, payload);
};

// Public
Handler.prototype.authenticate = function authenticate() {
  if(this._is_valid_app_key()) {
    this.connection.establish();
  } else {
    this.error({ code: 4001, message: "Could not find app by key " + this._app_key() });
    this.socket.close();
  }
};

Handler.prototype.on_message = function on_message(msg){
  msg = JSON.parse(msg);
  event = msg.event.replace(/^\/raptor\//, "raptor_");
  if(event == "client"){ // TODO: replace trailing/leading slashes
    msg.socket_id = this.connection.socket_id;
    this.channel.send_client_message(msg);
  } else if(event in this) {
    this[event](msg);
  }
};

Handler.prototype.raptor_subscribe = function raptor_subscribe(msg) {
  var channel_id = msg.data.channel;
  var klass = this._subscription_klass(channel_id);

  if(_.contains(this.subscriptions, channel_id)) {
    this.error({ code: nil, message: "Existing subscription to " + channel_id });
  } else {
    var new_subscription = new klass(this.connection.socket, this.connection.socket_id, msg);
    this.subscriptions[channel_id] = new_subscription.subscribe();
  }
};

// Private
Handler.prototype._is_valid_app_key = function _is_valid_app_key(){
  if(this._app_key() == '765ec374ae0a69f4ce44')
    return true;
  else
    return false;
};

Handler.prototype._app_key = function _app_key(){
  var app_key = this._path().split("/")[1];
  return app_key;
}

Handler.prototype._path = function _path(){
  var urlParts = urlParser.parse(this.handshake.target.url);
  return urlParts.pathname;
};

Handler.prototype._subscription_klass = function _subscription_klass(channel_id){
  var klass;

  var match = channel_id.match(/^\/(private|presence)\//);

  if(match !== null && (match[1] !== null)) {
    klass = require('./' + match[1] + '-subscription');
  }

  if(klass === undefined || klass === null){
    klass = require('./subscription');
  }

  return klass;
};

// Export
module.exports = Handler;
